Sources/SwiftCompilerPlugin/Macros.swift

```diff
@@ -0,0 +1,254 @@
+      let rewritten = try exprMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    if let declMacroDef = macroDef as? DeclarationMacro.Type {
+      let rewritten = try declMacroDef.expansion(of: macroSyntax, in: context)
+      let diagnostics = context.diagnostics.map {
+        PluginMessage.Diagnostic(from: $0, in: sourceManager)
+      }
+
+      let resultMessage = PluginToHostMessage.expandFreestandingMacroResult(
+        expandedSource: rewritten.description,
+        diagnostics: diagnostics
+      )
+
+      try pluginHostConnection.sendMessage(resultMessage)
+      return
+    }
+
+    let diag = PluginMessage.Diagnostic(
+      from: Diagnostic(node: syntax, message: MacroExpansionError.macroTypeNotFound),
+      in: sourceManager
+    )
+
+    try pluginHostConnection.sendMessage(
+      .expandFreestandingMacroResult(expandedSource: "", diagnostics: [diag])
+    )
+  }
+
+  /// Expand `@attached(XXX)` macros.
+  func expandAttachedMacro(
+    macro: PluginMessage.MacroReference,
+    macroRole: PluginMessage.MacroRole,
+    discriminator: String,
+    customAttributeSyntax: PluginMessage.Syntax,
+    declSyntax: PluginMessage.Syntax,
+    parentDeclSyntax: PluginMessage.Syntax?
+  ) throws {
+    let sourceManager = SourceManager()
+    let context = PluginMacroExpansionContext(
+      sourceManager: sourceManager,
+      discriminator: discriminator
+    )
+
+    guard let macroDefinition = resolveMacro(macro) else {
+      fatalError("macro type not found: \(macro.moduleName).\(macro.typeName)")
+    }
+
+    let customAttributeNode = sourceManager.add(customAttributeSyntax).cast(AttributeSyntax.self)
+    let declarationNode = sourceManager.add(declSyntax).cast(DeclSyntax.self)
+
+    let expanded: [String]
+    switch (macroDefinition, macroRole) {
+    case (let attachedMacro as AccessorMacro.Type, .accessor):
+      let accessors = try attachedMacro.expansion(
+        of: customAttributeNode,
+        providingAccessorsOf: declarationNode,
+        in: context
+      )
+      expanded =
+        accessors
+        .map { $0.trimmedDescription }
+
+    case (let attachedMacro as MemberAttributeMacro.Type, .memberAttribute):
+      guard
+        let parentDeclSyntax = parentDeclSyntax,
+        let parentDeclGroup = sourceManager.add(parentDeclSyntax).asProtocol(DeclGroupSyntax.self)
+      else {
+        fatalError("parentDecl is mandatory for MemberAttributeMacro")
+      }
+
+      // Local function to expand a member atribute macro once we've opened up
+      // the existential.
+      func expandMemberAttributeMacro<Node: DeclGroupSyntax>(
```

Comment:
Couldn’t the content of this just be a trailing closure to `_openExistential`? Same below.

Reply:
[Aligning with `ASTGen`](https://github.com/apple/swift/blob/a0e1810f627add51fbbe3ddf2c3ce9a9c6e3afdc/lib/ASTGen/Sources/ASTGen/Macros.swift#L677-L695) cc: @DougGregor